File: clients/eventodb-node/tests/last.test.ts
import { describe, test, expect, afterEach } from 'vitest';
import { setupTest, randomStreamName, type TestContext } from './helpers.js';

describe('LAST Tests', () => {
  const contexts: TestContext[] = [];

  afterEach(async () => {
    await Promise.all(contexts.map(ctx => ctx.cleanup()));
    contexts.length = 0;
  });

  test('LAST-001: Last message from non-empty stream', async () => {
    const ctx = await setupTest('last-001');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    // Write 5 messages
    for (let i = 0; i < 5; i++) {
      await ctx.client.streamWrite(stream, {
        type: 'Event',
        data: { n: i }
      });
    }

    const lastMessage = await ctx.client.streamLast(stream);
    expect(lastMessage).not.toBe(null);
    
    if (lastMessage) {
      const [, , position, , data] = lastMessage;
      expect(position).toBe(4); // Last position in 0-indexed stream
      expect(data).toEqual({ n: 4 });
    }
  });

  test('LAST-002: Last message from empty stream', async () => {
    const ctx = await setupTest('last-002');
    contexts.push(ctx);

    const stream = randomStreamName();
    const lastMessage = await ctx.client.streamLast(stream);
    
    expect(lastMessage).toBe(null);
  });

  test('LAST-003: Last message filtered by type', async () => {
    const ctx = await setupTest('last-003');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    // Write messages: TypeA, TypeB, TypeA, TypeB, TypeA
    await ctx.client.streamWrite(stream, { type: 'TypeA', data: { pos: 0 } });
    await ctx.client.streamWrite(stream, { type: 'TypeB', data: { pos: 1 } });
    await ctx.client.streamWrite(stream, { type: 'TypeA', data: { pos: 2 } });
    await ctx.client.streamWrite(stream, { type: 'TypeB', data: { pos: 3 } });
    await ctx.client.streamWrite(stream, { type: 'TypeA', data: { pos: 4 } });

    const lastTypeB = await ctx.client.streamLast(stream, { type: 'TypeB' });
    expect(lastTypeB).not.toBe(null);
    
    if (lastTypeB) {
      const [, type, position, , data] = lastTypeB;
      expect(type).toBe('TypeB');
      expect(position).toBe(3);
      expect(data).toEqual({ pos: 3 });
    }
  });

  test('LAST-004: Last message type filter no match', async () => {
    const ctx = await setupTest('last-004');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    // Write only TypeA messages
    await ctx.client.streamWrite(stream, { type: 'TypeA', data: { n: 1 } });
    await ctx.client.streamWrite(stream, { type: 'TypeA', data: { n: 2 } });

    const lastTypeB = await ctx.client.streamLast(stream, { type: 'TypeB' });
    expect(lastTypeB).toBe(null);
  });
});

File: clients/messagedb_ex/test/last_test.exs
defmodule MessagedbEx.LastTest do
  use ExUnit.Case, async: true
  import MessagedbEx.TestHelper

  setup do
    {client, namespace_id, _token} = create_test_namespace("last")

    on_exit(fn ->
      cleanup_namespace(client, namespace_id)
    end)

    %{client: client}
  end

  test "LAST-001: Last message from non-empty stream", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{}}

    # Write 5 messages
    client =
      Enum.reduce(1..5, client, fn _, c ->
        {:ok, _result, new_c} = MessagedbEx.stream_write(c, stream, message)
        new_c
      end)

    {:ok, last_msg, _client} = MessagedbEx.stream_last(client, stream)
    [_id, _type, position, _gpos, _data, _metadata, _time] = last_msg

    assert position == 4
  end

  test "LAST-002: Last message from empty stream", %{client: client} do
    stream = unique_stream()

    assert {:ok, nil, _client} = MessagedbEx.stream_last(client, stream)
  end

  test "LAST-003: Last message filtered by type", %{client: client} do
    stream = unique_stream()

    # Write messages: TypeA, TypeB, TypeA, TypeB, TypeA
    types = ["TypeA", "TypeB", "TypeA", "TypeB", "TypeA"]

    client =
      Enum.reduce(types, client, fn type, c ->
        message = %{type: type, data: %{}}
        {:ok, _result, new_c} = MessagedbEx.stream_write(c, stream, message)
        new_c
      end)

    {:ok, last_msg, _client} = MessagedbEx.stream_last(client, stream, %{type: "TypeB"})
    [_id, type, position, _gpos, _data, _metadata, _time] = last_msg

    assert type == "TypeB"
    assert position == 3
  end

  test "LAST-004: Last message type filter no match", %{client: client} do
    stream = unique_stream()

    # Write only TypeA messages
    client =
      Enum.reduce(1..3, client, fn _, c ->
        message = %{type: "TypeA", data: %{}}
        {:ok, _result, new_c} = MessagedbEx.stream_write(c, stream, message)
        new_c
      end)

    assert {:ok, nil, _client} = MessagedbEx.stream_last(client, stream, %{type: "TypeB"})
  end
end
